Olá, Gabriel!

Meu nome é Ramon e te ajudarei neste projeto. Fico feliz em rever seu projeto hoje.

Ao ao longo do texto farei algumas observações sobre melhorias no código e também farei comentários sobre suas percepções sobre o assunto. Estarei aberta a feedbacks e discussões sobre o tema.

Você encontrará meus comentários abaixo - **por favor, não os mova, modifique ou exclua**.

Você pode encontrar meus comentários em caixas verdes, amarelas ou vermelhas como esta:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>



<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Observações. Algumas recomendações.
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Precisa de correções. O bloqueio requer algumas correções. O trabalho não pode ser aceito com os comentários em vermelho.
</div>

Você pode me responder usando isso:

<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

# Recomendação um dos planos da Megaline: Smart ou Ultra.



A operadora de celular Megaline quer desenvolver um modelo que possa analisar o comportamento do cliente e recomendar um dos planos mais recentes da Megaline: Smart ou Ultra.

## Iniciação

Carregue bibliotecas que você acredita que são necessárias para o projeto. Você pode perceber depois que você precisa de bibliotecas adicionais conforme você avança, o que é totalmente normal - apenas certifique-se de atualizar essa seção quando perceber isso.

In [1]:
# Carregando todas as bibliotecas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats as st
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Carregue os Dados

In [2]:
# Carregue o arquivo com os dados em um DataFrame
df = pd.read_csv('/datasets/users_behavior_upd.csv')


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Explore os dados iniciais

**Descrição dos dados**
- `сalls` - (número de chamadas)
- `minutes` - (duração total da chamada em minutos)
- `messages` - (número de mensagens de texto)
- `mb_used` - (Tráfego de Internet usado em MB)
- `is_ultra` - (plano para o mês atual (Ultimate - 1, Surf - 0))

Agora vamos explorar nossos dados. Você vai querer ver quantas colunas e linhas ele tem, veja algumas linhas para verificar possíveis problemas com os dados.

In [3]:
# Vamos ver quantas linhas e colunas nosso conjunto de dados tem
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   calls        3214 non-null   float64
 1   minutes      3214 non-null   float64
 2   messages     3214 non-null   float64
 3   mb_used      3214 non-null   float64
 4   is_ultimate  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB



<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [4]:
# vamos exibir as primeiras 10 linhas
df.head(10)

calls  minutes  messages   mb_used  is_ultimate
0   40.0   311.90      83.0  19915.42            0
1   85.0   516.75      56.0  22696.96            0
2   77.0   467.66      86.0  21060.45            0
3  106.0   745.53      81.0   8437.39            1
4   66.0   418.74       1.0  14502.75            0
5   58.0   344.56      21.0  15823.37            0
6   57.0   431.64      20.0   3738.90            1
7   15.0   132.40       6.0  21911.60            0
8    7.0    43.39       3.0   2538.67            1
9   90.0   665.41      38.0  17358.61            0


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [5]:
# Verificar Valores Duplicados
df.duplicated().sum()

0


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [6]:
# Verificar Valores Ausentes
df.isna().sum()

calls          0
minutes        0
messages       0
mb_used        0
is_ultimate    0
dtype: int64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Transformação de dados
Vamos examinar cada coluna para ver quais problemas podemos ter nelas.

In [7]:
df['calls'].describe()

count    3214.000000
mean       63.038892
std        33.236368
min         0.000000
25%        40.000000
50%        62.000000
75%        82.000000
max       244.000000
Name: calls, dtype: float64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [8]:
df['minutes'].describe()

count    3214.000000
mean      438.208787
std       234.569872
min         0.000000
25%       274.575000
50%       430.600000
75%       571.927500
max      1632.060000
Name: minutes, dtype: float64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [9]:
df['messages'].describe()

count    3214.000000
mean       38.281269
std        36.148326
min         0.000000
25%         9.000000
50%        30.000000
75%        57.000000
max       224.000000
Name: messages, dtype: float64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [10]:
df['mb_used'].describe()

count     3214.000000
mean     17207.673836
std       7570.968246
min          0.000000
25%      12491.902500
50%      16943.235000
75%      21424.700000
max      49745.730000
Name: mb_used, dtype: float64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [11]:
df['is_ultimate'].value_counts(ascending=False,dropna=False)

0    2229
1     985
Name: is_ultimate, dtype: int64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Dividindo os dados de origem em um conjunto de treinamento, um conjunto de validação e um conjunto de teste.

In [12]:
df_train_valid, df_test = train_test_split(df, test_size=0.2)
df_train, df_valid = train_test_split(df_train_valid, test_size=0.25)


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [13]:
features_train = df_train.drop(['is_ultimate'], axis=1)
target_train = df_train['is_ultimate']
features_valid = df_valid.drop(['is_ultimate'], axis=1)
target_valid = df_valid['is_ultimate']
features_test = df_test.drop(['is_ultimate'], axis=1)
target_test = df_test['is_ultimate']


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

Separamos em 3 os dados: conjunto de treinamento, um conjunto de validação e um conjunto de teste.


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Melhorando a qualidade do modelo

In [14]:
for depth in range(1, 9):
    model_dctc = DecisionTreeClassifier(random_state=5432, max_depth=depth)
    model_dctc.fit(features_train, target_train)
    predictions_valid = model_dctc.predict(features_valid)
    print('max_depth =', depth, ': ', end='')
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 1 : 0.7480559875583204
max_depth = 2 : 0.80248833592535
max_depth = 3 : 0.8087091757387247
max_depth = 4 : 0.7947122861586314
max_depth = 5 : 0.8055987558320373
max_depth = 6 : 0.7838258164852255
max_depth = 7 : 0.8009331259720062
max_depth = 8 : 0.7962674961119751


Fizemos um For para saber quanto max_depth está a melhor acurácia, e o melhor foi o 4,
aplicaremos abaixo e procuraremos o melhor n_estimators pra o nosso modelo.


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto. Usou treino para treinar e validação para avaliar o resultado a fim de calibrar hiperpãmetro depth
</div>

In [15]:
best_score = 0
best_est = 0
for est in range(1, 11):
    model_best = RandomForestClassifier(random_state=5432,max_depth=4, n_estimators=est)
    model_best.fit(features_train, target_train)
    score = model_best.score(features_valid, target_valid)
    if score > best_score:
        best_score = score
        best_est = est

print("A acurácia do melhor modelo no conjunto de validação (n_estimators = {}): {}".format(best_est, best_score))

A acurácia do melhor modelo no conjunto de validação (n_estimators = 7): 0.8164852255054432



<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto. Porém, você realizou a busca de forma independente, isto é, sem levar em consideração a interação entre os parâmetros. O ideal é usar grid search https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html
</div>

n_estimators = 6 foi o queteve melhor acurácia


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [16]:
model = RandomForestClassifier(random_state=5432,max_depth=4, n_estimators=6)
model.fit(features_train,target_train)
model_valid = model.predict(features_valid)


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [18]:
accuracy_score(target_valid, model_valid)

0.807153965785381

0.7962674961119751 foi a acurácia com melhor desepenho no nosso modelo.Agora passaremos para o conjunto de teste.


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [19]:
final_model = RandomForestClassifier(random_state=5432,max_depth=4, n_estimators=6)
final_model.fit(features_valid,target_valid)
predictions_test = final_model.predict(features_test)


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [20]:
accuracy_score(target_test,predictions_test)

0.7744945567651633


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

Criamos um modelo com 0.78 de acurácia para a escolha da recomendações do plano Ultra ou Smart.


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Conclusão

Neste projeto, importamos bibliotecas como "pandas" e "numpy" que habitualmente já utilizamos, e sklearn para criação do nosso modelo.

Verificamos se realmente o pre-processamento foi feito, para que não tivesse acontecido algum problema na criação do nosso modelo.


E desenvolvemos um modelo para recomedação, e usamos a floresta aleatória pois sua acurácia é alta mediante a treinamento dos modelos.


<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho. Você apresentou projeto correto, limpo. Única sugestão é quando for calibrar hiperparâmetro usar estratégia de Grid Search (demora mais mas considera a interação entre os hiperparâmetros).
    
Bom trabalho.
</div>